hpccm/building_blocks/scif.py

```diff
@@ -0,0 +1,122 @@
+# pylint: disable=invalid-name, too-few-public-methods
+# pylint: disable=too-many-instance-attributes
+
+"""SCIF building block"""
+
+from __future__ import absolute_import
+from __future__ import unicode_literals
+from __future__ import print_function
+
+from hpccm.building_blocks.packages import packages
+from hpccm.common import container_type
+from hpccm.primitives.comment import comment
+from hpccm.primitives.copy import copy
+from hpccm.primitives.runscript import runscript
+from hpccm.primitives.shell import shell
+
+import hpccm.config
+import logging
+import os
+
+class scif(object):
+    """SCIF building block"""
+    initialized = False
+
+    def __init__(self, **kwargs):
+        """Initialize building block"""
+        self.name = kwargs.get('name', '')
+        self.scif_version = kwargs.get('scif_version', '0.0.76')
+        self.__layers = []
+        self.__separator = kwargs.get('separator', '\n\n')
+
+        self.__entrypoint = kwargs.get('entrypoint', False)
+        self.instructions = [comment("SCIF app {0}".format(self.name))]
+        if hpccm.config.g_ctype == container_type.DOCKER and not scif.initialized:
+            # Install the scif-tool on first scif app
+            self.instructions.extend([
+                comment("Begin SCI-F installation"),
+                packages(ospackages=['python-pip', 'python-setuptools']),
+                shell(commands=[
+                  'pip install wheel',
+                  'pip install scif=={0}'.format(self.scif_version)
+                ]),
+                comment("End SCI-F installation")
+            ])
+            scif.initialized = True
+
+        allowed_primitives = ["comment", "copy", "environment", "help",
+                              "label", "runscript", "shell", "test"]
+        self.__primitive_status = {i: False for i in allowed_primitives}
+
+    def __iadd__(self, layer):
+        """Add the scif layer.  Allows "+=" syntax."""
+        if isinstance(layer, list):
+            for l in layer:
+                self.check_layer(l)
+            self.__layers.extend(layer)
+        else:
+            self.check_layer(layer)
+            self.__layers.append(layer)
+        return self
+
+    def check_layer(self, layer):
+        class_name = type(layer).__name__
+        if class_name not in self.__primitive_status:
+            logging.exception(
+                'You cannot add `{0}` to a scif! Use primitives which support '
+                'the _app-parameter! I.e.: {1}'.format(class_name,
+                    str(self.__primitive_status.keys())))
+
+        if self.__primitive_status[class_name]:
+            logging.exception(
+                'Duplicate `{0}` primitive given for scif app. '
+                'Each primitive is allowed just once!'.format(class_name))
+            raise()  # TODO: the logging exception is not raised?!
```

Comment:
Not a big deal, but the exception stack traces that get printed out are a little ugly.  Use `logging.error()` / `raise` as used elsewhere?

Also a TODO.